packages/common/src/tools/queue/UniqueTaskQueue.ts

```diff
@@ -0,0 +1,39 @@
+import { Logger } from '../Logger'
+import { TaskQueue } from './TaskQueue'
+import { TaskQueueOpts } from './types'
+
+export class UniqueTaskQueue<T extends string | number> extends TaskQueue<T> {
+  private taskSet: Set<T> = new Set()
+
+  constructor(
+    executeTask: (task: T) => Promise<void>,
+    logger: Logger,
+    opts?: TaskQueueOpts<T>,
+  ) {
+    const wrappedExecute = async (task: T) => {
+      await executeTask(task)
+      this.taskSet.delete(task)
+    }
+
+    super(wrappedExecute, logger, opts)
+  }
+
+  override addIfEmpty(task: T): void {
+    this.addUnique(task, super.addIfEmpty.bind(this))
+  }
+
+  override addToBack(task: T): void {
+    this.addUnique(task, super.addToBack.bind(this))
+  }
+
+  override addToFront(task: T): void {
+    this.addUnique(task, super.addToFront.bind(this))
+  }
+
+  private addUnique(task: T, method: (task: T) => void) {
+    if (!this.taskSet.has(task)) {
+      this.taskSet.add(task)
+      method(task)
+    }
+  }
```

Comment:
Since this is already merged I am only leaving this comment as a consideration for the future. Instead of rebinding the method every time one could just make the if impure

```ts
  override addToFront(task: T): void {
    if (this.shouldAdd(task)) {
      super.addToFront(task)
    }
  }

  private shouldAdd(task: T) {
    if (!this.taskSet.has(task)) {
      this.taskSet.add(task)
      return true
    }
    return false
  }
```